### DISCUSSION TITLE
Codes cut off irregularly when parsing a np.array of SMILES codes

### DISCUSSION THREAD

**Discussion by gloganbart** [rdkit/rdkit#6077] [Q&A] [answered]
Hey there!

I'm writing a script that requires I load in a library of SMILES strings from a tab-delimited .txt file (exported from Reaxys) as a numpy array.

I used a small (~400) set of SMILES strings in development, and no problems occurred. Each SMILES string was parsed properly and converted to a MOL as follows:

```
smiles_codes = numpy.loadtxt('400_smiles_codes_from_reaxys.txt', 
                         delimiter = '',
                         dtype = str)

smiles_to_mol = numpy.vectorize(Chem.MolFromSmiles)
mol_codes = smiles_to_mol(smiles_codes)
```

However, when I exported a larger (~15000) set of SMILES strings from Reaxys and ran the same code on the new dataset, the first ~130 of the SMILES strings were properly parsed before the following error arose in printing the numpy array of MOLs:

```
smiles_codes = numpy.loadtxt('15000_smiles_codes_from_reaxys.txt', 
                         delimiter = '',
                         dtype = str)

smiles_to_mol = numpy.vectorize(Chem.MolFromSmiles)
mol_codes = smiles_to_mol(smiles_codes)

print(mol_codes
```

This code results in the following:
```
 [<rdkit.Chem.rdchem.Mol object at 0x7f9fb1d8b6d0>
<rdkit.Chem.rdchem.Mol object at 0x7f9fb1d8b740>
<rdkit.Chem.rdchem.Mol object at 0x7f9fb1d8b7b0> ...
<rdkit.Chem.rdchem.Mol object at 0x7f9fb01a0cf0>
<rdkit.Chem.rdchem.Mol object at 0x7f9fb01a0d60>
<rdkit.Chem.rdchem.Mol object at 0x7f9fb01a0dd0>]
[01:02:32] SMILES Parse Error: extra open parentheses for input: 'CSC1=NC=C(C'
```

Somehow, RDkit is "cutting off" certain SMILES strings. 

When I go look in the .txt for 'CSC1=NC=C(C', I find instead 'CSC1=NC=C(C#N)C(N)=N1'.

It seems other SMILES strings with the # character cause the same issue.

Does anyone have an explanation for this? thanks!

**Comment by greglandrum** [comment]
I guess the numpy reader is treating the # character as a comment delimiter.

**Comment by apahl** [comment, marked as answer]
Yes, it would appear so.  
This is the [function signature](https://numpy.org/doc/stable/reference/generated/numpy.loadtxt.html):

```
numpy.loadtxt(fname, dtype=<class 'float'>, comments='#', delimiter=None, converters=None, skiprows=0, usecols=None, unpack=False, ndmin=0, encoding='bytes', max_rows=None, *, quotechar=None, like=None)
```
Setting the `comments` option to `None` should solve the problem.

**Comment by gloganbart** [comment]
Thanks everyone for your help!
